In [1]:
import pandas as pd

df = pd.read_csv('Result 2025-09-23 05-01-51.csv')
df

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,stories,fcc_asr_number,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason
0,942614,Phoenix Tower International,NaN,Phoenix Tower International,2019-08-22 20:53:42.284115,2023-01-13 15:25:12.493704,33.94619,-118.28200,CA12409,160407,...,NaN,NaN,NaN,NaN,NaN,"FIGUEROA ST WL 75F S OF 99TH ST S/N , August F...",NaN,No,Unconfirmed,Focus Asset
1,942614,Phoenix Tower International,943927.0,Phoenix Tower International,2019-08-22 21:36:14.93064,2023-06-26 12:05:50.800659,33.94574,-118.28200,MAIN ST EL 185F N OF 110TH ST SF,161773,...,NaN,NaN,NaN,NaN,NaN,"MAIN ST EL 185F N OF 110TH ST SF , , CA 90003,...",NaN,No,Unconfirmed,Proximity Audit (48m)
2,942664,Phoenix Tower International,NaN,Phoenix Tower International,2019-08-22 20:55:21.34445,2023-01-13 11:49:35.586868,34.03929,-118.19900,CA12465,160463,...,NaN,NaN,NaN,NaN,NaN,"1ST ST SL 2F E OF DACOTAH ST E/W , Los Angeles...",NaN,No,Unconfirmed,Focus Asset
3,942664,Phoenix Tower International,942203.0,Phoenix Tower International,2019-08-22 20:40:05.047291,2023-01-13 13:04:52.922687,34.03910,-118.19900,CA11987,159985,...,NaN,NaN,NaN,NaN,NaN,"1ST ST SL 53F W OF FRESNO ST WF , Los Angeles,...",NaN,No,Unconfirmed,Proximity Audit (21m)
4,942761,Phoenix Tower International,NaN,Phoenix Tower International,2019-08-22 20:58:33.921034,2023-01-13 12:41:13.613007,33.98899,-118.31900,CA12570,160568,...,NaN,NaN,NaN,NaN,NaN,"SLAUSON AVE NL 42F E OF 3RD AVE EF , Los Angel...",NaN,No,Unconfirmed,Focus Asset
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
64688,1161828,ASR,264885.0,Crown Castle,2015-04-20 20:26:56.595238,2017-07-28 23:55:23.576292,39.00210,-95.73923,SKYLINE PARK/BURNETT'S MOUND,877829,...,NaN,NaN,NaN,NaN,NaN,"3511 SW SKYLINE PARKWAY , TOPEKA, KS 66614, UN...",2007-01-12,No,Active,Proximity Audit (7m)
64689,1161829,ASR,NaN,ASR,2025-04-01 15:38:18.798044,2025-04-09 11:29:48.458842,32.81861,-86.61714,NaN,NaN,...,NaN,1330244.0,2024-ASO-23463-OE,NaN,NaN,"Off Logan Road , Clanton, AL 35045, UNITED STATES",2024-11-26,No,Active,Focus Asset
64690,1161829,ASR,1136163.0,ASR,2024-01-23 15:24:06.649527,2024-01-23 09:31:32.771277,32.81861,-86.61714,NaN,NaN,...,NaN,1327054.0,2023-ASO-27414-OE,NaN,NaN,"Off Logan Road , Clanton, AL 35046, UNITED STATES",2024-11-26,No,Active,Proximity Audit (0m)
64691,1162574,ASR,NaN,ASR,2025-07-24 21:02:34.966713,2025-07-28 09:49:51.734224,42.34739,-91.47869,NaN,NaN,...,NaN,1330947.0,2024-ACE-6738-OE,NaN,NaN,"3048 Hwy 13 IA-5254 , Ryan, IA 52330, UNITED S...",NaN,No,ASR-Granted,Focus Asset


In [2]:
df.dtypes

focus_asset_id           int64
focus_asset             object
associated_asset_id    float64
source                  object
created_at              object
updated_at              object
latitude               float64
longitude              float64
name                    object
operator_site_id        object
type                    object
description             object
operator_name           object
manager_name            object
fcc_owner_name          object
agl                    float64
amsl                   float64
ground_elevation       float64
haat                   float64
shelter                 object
power                   object
stories                float64
fcc_asr_number         float64
faa_study_number        object
cdbs_facility_id       float64
region                  object
address                 object
construction_date       object
stealth                 object
asset_status            object
audit_reason            object
dtype: object

In [3]:
df['focus_asset_id'] = df['focus_asset_id'].astype('Int64')
df['focus_asset'] = df['focus_asset'].astype('string')
df['associated_asset_id'] = df['associated_asset_id'].astype('Int64')
df['source'] = df['source'].astype('string')
df['name'] = df['name'].astype('string')
df['operator_site_id'] = df['operator_site_id'].astype('string')
df['type'] = df['type'].astype('string')
df['description'] = df['description'].astype('string')
df['operator_name'] = df['operator_name'].astype('string')
df['manager_name'] = df['manager_name'].astype('string')
df['fcc_owner_name'] = df['fcc_owner_name'].astype('string')
df['shelter'] = df['shelter'].astype('string')
df['power'] = df['power'].astype('string')
df['fcc_asr_number'] = df['fcc_asr_number'].astype('string')
df['faa_study_number'] = df['faa_study_number'].astype('string')
df['cdbs_facility_id'] = df['cdbs_facility_id'].astype('string')
df['region'] = df['region'].astype('string')
df['address'] = df['address'].astype('string')
df['stealth'] = df['stealth'].astype('string')
df['asset_status'] = df['asset_status'].astype('string')
df['audit_reason'] = df['audit_reason'].astype('string')

Let's calculate first the difference of distances of the associated tower assets from the focus/reference tower asset.

In [4]:
from geopy.distance import geodesic
import numpy as np

def calculate_distances_to_reference(df):
    df_copy = df.copy()
    df_copy['distance_to_reference'] = np.nan

    for group_name, group_df in df_copy.groupby('focus_asset_id'):
        reference_record = group_df[group_df['associated_asset_id'].isna()]

        if not reference_record.empty:
            ref_lat = reference_record['latitude'].iloc[0]
            ref_lon = reference_record['longitude'].iloc[0]
            reference_coords = (ref_lat, ref_lon)

            for index, row in group_df.iterrows():
                if pd.notna(row['associated_asset_id']):
                    record_coords = (row['latitude'], row['longitude'])
                    distance = geodesic(reference_coords, record_coords).meters
                    df_copy.loc[index, 'distance_to_reference'] = distance
    return df_copy

df_with_distances = calculate_distances_to_reference(df)
df_with_distances

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,fcc_asr_number,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference
0,942614,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:53:42.284115,2023-01-13 15:25:12.493704,33.94619,-118.28200,CA12409,160407,...,<NA>,<NA>,<NA>,<NA>,"FIGUEROA ST WL 75F S OF 99TH ST S/N , August F...",NaN,No,Unconfirmed,Focus Asset,NaN
1,942614,Phoenix Tower International,943927,Phoenix Tower International,2019-08-22 21:36:14.93064,2023-06-26 12:05:50.800659,33.94574,-118.28200,MAIN ST EL 185F N OF 110TH ST SF,161773,...,<NA>,<NA>,<NA>,<NA>,"MAIN ST EL 185F N OF 110TH ST SF , , CA 90003,...",NaN,No,Unconfirmed,Proximity Audit (48m),49.914635
2,942664,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:55:21.34445,2023-01-13 11:49:35.586868,34.03929,-118.19900,CA12465,160463,...,<NA>,<NA>,<NA>,<NA>,"1ST ST SL 2F E OF DACOTAH ST E/W , Los Angeles...",NaN,No,Unconfirmed,Focus Asset,NaN
3,942664,Phoenix Tower International,942203,Phoenix Tower International,2019-08-22 20:40:05.047291,2023-01-13 13:04:52.922687,34.03910,-118.19900,CA11987,159985,...,<NA>,<NA>,<NA>,<NA>,"1ST ST SL 53F W OF FRESNO ST WF , Los Angeles,...",NaN,No,Unconfirmed,Proximity Audit (21m),21.075388
4,942761,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:58:33.921034,2023-01-13 12:41:13.613007,33.98899,-118.31900,CA12570,160568,...,<NA>,<NA>,<NA>,<NA>,"SLAUSON AVE NL 42F E OF 3RD AVE EF , Los Angel...",NaN,No,Unconfirmed,Focus Asset,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
64688,1161828,ASR,264885,Crown Castle,2015-04-20 20:26:56.595238,2017-07-28 23:55:23.576292,39.00210,-95.73923,SKYLINE PARK/BURNETT'S MOUND,877829,...,<NA>,<NA>,<NA>,<NA>,"3511 SW SKYLINE PARKWAY , TOPEKA, KS 66614, UN...",2007-01-12,No,Active,Proximity Audit (7m),7.276909
64689,1161829,ASR,<NA>,ASR,2025-04-01 15:38:18.798044,2025-04-09 11:29:48.458842,32.81861,-86.61714,<NA>,<NA>,...,1330244.0,2024-ASO-23463-OE,<NA>,<NA>,"Off Logan Road , Clanton, AL 35045, UNITED STATES",2024-11-26,No,Active,Focus Asset,NaN
64690,1161829,ASR,1136163,ASR,2024-01-23 15:24:06.649527,2024-01-23 09:31:32.771277,32.81861,-86.61714,<NA>,<NA>,...,1327054.0,2023-ASO-27414-OE,<NA>,<NA>,"Off Logan Road , Clanton, AL 35046, UNITED STATES",2024-11-26,No,Active,Proximity Audit (0m),0.000000
64691,1162574,ASR,<NA>,ASR,2025-07-24 21:02:34.966713,2025-07-28 09:49:51.734224,42.34739,-91.47869,<NA>,<NA>,...,1330947.0,2024-ACE-6738-OE,<NA>,<NA>,"3048 Hwy 13 IA-5254 , Ryan, IA 52330, UNITED S...",NaN,No,ASR-Granted,Focus Asset,NaN


Now, let's find the unique Audit Reasons. Note that for each grouping, we can have multiple associated asset IDs. Thus, each associated asset IDs will have their own audit reason based on the reference/focus asset. Note that the "Focus asset" showing below is the Focus/Reference tower.

In [5]:
audit_reasons = df_with_distances['audit_reason'].value_counts()
print(audit_reasons)

audit_reason
Focus Asset                       28166
Proximity Audit (1m)              11226
Proximity Audit (0m)               6917
Manual Edit                        2397
Multiple matches                   1147
Proximity Audit (2m)                512
Proximity Audit (22m)               486
Proximity Audit (11m)               463
Proximity Audit (44m)               433
Proximity Audit (33m)               411
Proximity Audit (49m)               399
Proximity Audit (45m)               382
Proximity Audit (46m)               377
Proximity Audit (48m)               371
Proximity Audit (47m)               365
Proximity Audit (43m)               354
Proximity Audit (32m)               345
Proximity Audit (39m)               345
Proximity Audit (35m)               345
Proximity Audit (31m)               345
Proximity Audit (10m)               344
Proximity Audit (38m)               342
Proximity Audit (40m)               339
Proximity Audit (12m)               324
Proximity Audit (36m)      

We can see that almost all of the prox audits are coming from the "Proximity Audit" reason. These are towers that are within the vicinity of the reference/focus asset based on the radial distance set for proximity. NOTE: There's a specific matching logic on how the to tag the audit reason for each associated assets with respect to focus asset. I'll not go through with that here.

In [6]:
df_with_distances['distance_to_reference'].describe()

count    36527.000000
mean        11.857108
std         22.535720
min          0.000000
25%          0.934227
50%          1.392879
75%         22.307786
max       3029.144797
Name: distance_to_reference, dtype: float64

In [7]:
group_with_max_distance = df_with_distances.loc[df_with_distances['distance_to_reference'].round(decimals=2) == 3029.14]
group_with_max_distance

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,fcc_asr_number,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference
45665,1152648,SBA,1021930,sgaither,2021-02-03 21:51:51.099288,2025-04-16 21:10:02.998613,37.27886,-97.3915,Wellington II,KS-0011,...,1318490.0,2020-ACE-8306-OE,<NA>,<NA>,"1323 N A ST. , WELLINGTON, KS 67152, UNITED ST...",2022-03-08,No,Active,Proximity Audit (1m),3029.144797


In [8]:
df_with_distances[df_with_distances['focus_asset_id'] == 1152648]

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,fcc_asr_number,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference
45663,1152648,SBA,<NA>,SBA,2025-02-19 23:58:14.764593,2025-02-19 23:58:14.706249,37.26622,-97.42177,Wellington 3 KS,KS11395-A,...,<NA>,<NA>,<NA>,<NA>,"1602 W 8th Street , Wellington, KS 67152, UNIT...",NaN,No,Active,Focus Asset,NaN
45664,1152648,SBA,123580,ASR,2015-02-03 20:22:52.13941,2019-03-20 15:05:01.744757,37.26622,-97.42178,"Wellington 3, KS",KS11395-A,...,1257460.0,2008-ACE-2385-OE,<NA>,<NA>,", Wellington, KS 67152, UNITED STATES",2007-04-15,No,Active,Proximity Audit (0m),0.887004
45665,1152648,SBA,1021930,sgaither,2021-02-03 21:51:51.099288,2025-04-16 21:10:02.998613,37.27886,-97.39150,Wellington II,KS-0011,...,1318490.0,2020-ACE-8306-OE,<NA>,<NA>,"1323 N A ST. , WELLINGTON, KS 67152, UNITED ST...",2022-03-08,No,Active,Proximity Audit (1m),3029.144797


Why does the 3,029.144797m distance away from the reference tower is showing as Proximity Audit (1m)?

Let's find the difference in AGL of the reference tower against each associated tower.

In [9]:
def calculate_agldiff_to_reference(df):
    df_copy = df_with_distances.copy()
    df_copy['agldiff_to_reference'] = np.nan

    for group_name, group_df in df_copy.groupby('focus_asset_id'):
        reference_record = group_df[group_df['associated_asset_id'].isna()]

        if not reference_record.empty:
            ref_agl = reference_record['agl'].iloc[0]

            for index, row in group_df.iterrows():
                if pd.notna(row['associated_asset_id']):
                    record_agl = row['agl']
                    distance = ref_agl - record_agl
                    df_copy.loc[index, 'agldiff_to_reference'] = distance
    return df_copy

df_with_differences = calculate_agldiff_to_reference(df_with_distances)
df_with_differences

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,942614,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:53:42.284115,2023-01-13 15:25:12.493704,33.94619,-118.28200,CA12409,160407,...,<NA>,<NA>,<NA>,"FIGUEROA ST WL 75F S OF 99TH ST S/N , August F...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
1,942614,Phoenix Tower International,943927,Phoenix Tower International,2019-08-22 21:36:14.93064,2023-06-26 12:05:50.800659,33.94574,-118.28200,MAIN ST EL 185F N OF 110TH ST SF,161773,...,<NA>,<NA>,<NA>,"MAIN ST EL 185F N OF 110TH ST SF , , CA 90003,...",NaN,No,Unconfirmed,Proximity Audit (48m),49.914635,0.0
2,942664,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:55:21.34445,2023-01-13 11:49:35.586868,34.03929,-118.19900,CA12465,160463,...,<NA>,<NA>,<NA>,"1ST ST SL 2F E OF DACOTAH ST E/W , Los Angeles...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
3,942664,Phoenix Tower International,942203,Phoenix Tower International,2019-08-22 20:40:05.047291,2023-01-13 13:04:52.922687,34.03910,-118.19900,CA11987,159985,...,<NA>,<NA>,<NA>,"1ST ST SL 53F W OF FRESNO ST WF , Los Angeles,...",NaN,No,Unconfirmed,Proximity Audit (21m),21.075388,0.0
4,942761,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:58:33.921034,2023-01-13 12:41:13.613007,33.98899,-118.31900,CA12570,160568,...,<NA>,<NA>,<NA>,"SLAUSON AVE NL 42F E OF 3RD AVE EF , Los Angel...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
64688,1161828,ASR,264885,Crown Castle,2015-04-20 20:26:56.595238,2017-07-28 23:55:23.576292,39.00210,-95.73923,SKYLINE PARK/BURNETT'S MOUND,877829,...,<NA>,<NA>,<NA>,"3511 SW SKYLINE PARKWAY , TOPEKA, KS 66614, UN...",2007-01-12,No,Active,Proximity Audit (7m),7.276909,4.5
64689,1161829,ASR,<NA>,ASR,2025-04-01 15:38:18.798044,2025-04-09 11:29:48.458842,32.81861,-86.61714,<NA>,<NA>,...,2024-ASO-23463-OE,<NA>,<NA>,"Off Logan Road , Clanton, AL 35045, UNITED STATES",2024-11-26,No,Active,Focus Asset,NaN,NaN
64690,1161829,ASR,1136163,ASR,2024-01-23 15:24:06.649527,2024-01-23 09:31:32.771277,32.81861,-86.61714,<NA>,<NA>,...,2023-ASO-27414-OE,<NA>,<NA>,"Off Logan Road , Clanton, AL 35046, UNITED STATES",2024-11-26,No,Active,Proximity Audit (0m),0.000000,-20.4
64691,1162574,ASR,<NA>,ASR,2025-07-24 21:02:34.966713,2025-07-28 09:49:51.734224,42.34739,-91.47869,<NA>,<NA>,...,2024-ACE-6738-OE,<NA>,<NA>,"3048 Hwy 13 IA-5254 , Ryan, IA 52330, UNITED S...",NaN,No,ASR-Granted,Focus Asset,NaN,NaN


In [10]:
df_with_differences['agldiff_to_reference'].describe()

count    18945.000000
mean        -0.383363
std         13.066281
min       -758.800000
25%         -0.100000
50%          0.000000
75%          0.000000
max        340.600000
Name: agldiff_to_reference, dtype: float64

In [11]:
agl_differences = df_with_differences['agldiff_to_reference'].value_counts()
print(agl_differences)

agldiff_to_reference
 0.0     8432
-0.1     2651
-0.1      802
-0.1      547
-0.1      426
         ... 
-33.9       1
 67.7       1
-0.6        1
 28.1       1
-20.4       1
Name: count, Length: 1263, dtype: int64


---

Let's try to see how many groupings will fall under the <b>Case 1</b>.

<strong>Case 1: Both the focus asset (reference or top data in Sherlock’s prox audit) and associated asset/s (bottom data  in Sherlock’s prox audit) have the same FCC ASR and FAA registry number</strong> <br>

In [12]:
df_with_differences

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,942614,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:53:42.284115,2023-01-13 15:25:12.493704,33.94619,-118.28200,CA12409,160407,...,<NA>,<NA>,<NA>,"FIGUEROA ST WL 75F S OF 99TH ST S/N , August F...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
1,942614,Phoenix Tower International,943927,Phoenix Tower International,2019-08-22 21:36:14.93064,2023-06-26 12:05:50.800659,33.94574,-118.28200,MAIN ST EL 185F N OF 110TH ST SF,161773,...,<NA>,<NA>,<NA>,"MAIN ST EL 185F N OF 110TH ST SF , , CA 90003,...",NaN,No,Unconfirmed,Proximity Audit (48m),49.914635,0.0
2,942664,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:55:21.34445,2023-01-13 11:49:35.586868,34.03929,-118.19900,CA12465,160463,...,<NA>,<NA>,<NA>,"1ST ST SL 2F E OF DACOTAH ST E/W , Los Angeles...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
3,942664,Phoenix Tower International,942203,Phoenix Tower International,2019-08-22 20:40:05.047291,2023-01-13 13:04:52.922687,34.03910,-118.19900,CA11987,159985,...,<NA>,<NA>,<NA>,"1ST ST SL 53F W OF FRESNO ST WF , Los Angeles,...",NaN,No,Unconfirmed,Proximity Audit (21m),21.075388,0.0
4,942761,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:58:33.921034,2023-01-13 12:41:13.613007,33.98899,-118.31900,CA12570,160568,...,<NA>,<NA>,<NA>,"SLAUSON AVE NL 42F E OF 3RD AVE EF , Los Angel...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
64688,1161828,ASR,264885,Crown Castle,2015-04-20 20:26:56.595238,2017-07-28 23:55:23.576292,39.00210,-95.73923,SKYLINE PARK/BURNETT'S MOUND,877829,...,<NA>,<NA>,<NA>,"3511 SW SKYLINE PARKWAY , TOPEKA, KS 66614, UN...",2007-01-12,No,Active,Proximity Audit (7m),7.276909,4.5
64689,1161829,ASR,<NA>,ASR,2025-04-01 15:38:18.798044,2025-04-09 11:29:48.458842,32.81861,-86.61714,<NA>,<NA>,...,2024-ASO-23463-OE,<NA>,<NA>,"Off Logan Road , Clanton, AL 35045, UNITED STATES",2024-11-26,No,Active,Focus Asset,NaN,NaN
64690,1161829,ASR,1136163,ASR,2024-01-23 15:24:06.649527,2024-01-23 09:31:32.771277,32.81861,-86.61714,<NA>,<NA>,...,2023-ASO-27414-OE,<NA>,<NA>,"Off Logan Road , Clanton, AL 35046, UNITED STATES",2024-11-26,No,Active,Proximity Audit (0m),0.000000,-20.4
64691,1162574,ASR,<NA>,ASR,2025-07-24 21:02:34.966713,2025-07-28 09:49:51.734224,42.34739,-91.47869,<NA>,<NA>,...,2024-ACE-6738-OE,<NA>,<NA>,"3048 Hwy 13 IA-5254 , Ryan, IA 52330, UNITED S...",NaN,No,ASR-Granted,Focus Asset,NaN,NaN


In [13]:
column_names_list = list(df_with_differences.columns)
column_names_list

['focus_asset_id',
 'focus_asset',
 'associated_asset_id',
 'source',
 'created_at',
 'updated_at',
 'latitude',
 'longitude',
 'name',
 'operator_site_id',
 'type',
 'description',
 'operator_name',
 'manager_name',
 'fcc_owner_name',
 'agl',
 'amsl',
 'ground_elevation',
 'haat',
 'shelter',
 'power',
 'stories',
 'fcc_asr_number',
 'faa_study_number',
 'cdbs_facility_id',
 'region',
 'address',
 'construction_date',
 'stealth',
 'asset_status',
 'audit_reason',
 'distance_to_reference',
 'agldiff_to_reference']

In [14]:
def find_case_1_candidates(df_with_differences):

    case_1_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']

            if pd.notnull(ref_fcc) and pd.notnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'] == ref_fcc) &
                    (paired_records_df['faa_study_number'] == ref_faa) &
                    (paired_records_df['fcc_asr_number'].notnull()) &
                    (paired_records_df['faa_study_number'].notnull())
                ]

                if not matching_paired_records.empty:
                    case_1_records.append(reference_record_df)
                    case_1_records.append(matching_paired_records)

    if case_1_records:
        case_1_candidates = pd.concat(case_1_records, ignore_index=True)
        return case_1_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_1_candidates = find_case_1_candidates(df_with_differences)
case_1_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1139789,ASR,<NA>,ASR,2021-02-22 14:55:46.271143,2024-07-16 10:20:53.166299,43.19161,-86.09733,<NA>,<NA>,...,2020-AGL-18544-OE,<NA>,<NA>,"5998 Heights Ravenna Road (MI-0028) , Fruitpor...",2020-06-25,No,Active,Focus Asset,NaN,NaN
1,1139789,ASR,994367,ASR,2020-05-13 12:39:45.368753,2024-06-25 09:36:13.925015,43.19161,-86.09733,Heights Ravenna Rd,MI-0028,...,2020-AGL-18544-OE,<NA>,<NA>,"5998 Heights Ravenna Rd , Fruitport, MI 49415,...",2020-06-25,No,In development,ASR Operator Mismatch,0.000000,0.7
2,1139876,ASR,<NA>,ASR,2021-02-22 14:55:46.271143,2024-07-23 10:08:33.984079,43.19161,-86.09733,<NA>,<NA>,...,2020-AGL-18544-OE,<NA>,<NA>,"5998 Heights Ravenna Road (MI-0028) , Fruitpor...",2020-06-25,No,Active,Focus Asset,NaN,NaN
3,1139876,ASR,994367,ASR,2020-05-13 12:39:45.368753,2024-06-25 09:36:13.925015,43.19161,-86.09733,Heights Ravenna Rd,MI-0028,...,2020-AGL-18544-OE,<NA>,<NA>,"5998 Heights Ravenna Rd , Fruitport, MI 49415,...",2020-06-25,No,In development,ASR Operator Mismatch,0.000000,0.7
4,1139926,ASR,<NA>,ASR,2021-02-22 14:55:46.271143,2024-07-30 11:28:32.685619,43.19161,-86.09733,<NA>,<NA>,...,2020-AGL-18544-OE,<NA>,<NA>,"5998 Heights Ravenna Road (MI-0028) , Fruitpor...",2020-06-25,No,Active,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
160,1147155,Everest Infrastructure Partners,833422,dledbetter,2018-09-20 10:16:32.231523,2018-09-21 14:49:34.622583,36.04094,-79.22089,Buckhhorn,00060,...,2015-ASO-2179-OE,<NA>,<NA>,"2507 Buckhorn Road , Mebane, NC 27302, UNITED ...",2015-05-19,No,Active,Proximity Audit (0m),0.901171,-0.1
161,1147156,Everest Infrastructure Partners,<NA>,Everest Infrastructure Partners,2025-02-18 16:27:48.0568,2025-02-18 16:27:47.993715,35.40977,-77.98728,Goldsboro,US596436,...,2020-ASO-23207-OE,<NA>,<NA>,"209z Hooks River Rd , Goldsboro, NC 27530, UNI...",NaN,No,Active,Focus Asset,NaN,NaN
162,1147156,Everest Infrastructure Partners,1137554,sgaither,2015-02-03 23:55:20.061725,2024-06-13 15:43:04.18398,35.40977,-77.98728,Goldsboro,US596436,...,2020-ASO-23207-OE,<NA>,<NA>,"209 Hooks River Rd , GOLDSBORO, NC 27530, UNIT...",2015-04-19,No,Active,Manual Edit,0.000000,-0.1
163,1147158,Everest Infrastructure Partners,<NA>,Everest Infrastructure Partners,2025-02-18 16:27:57.187729,2025-02-18 16:27:57.124848,34.03211,-86.31711,ALTOONA - DWNTWN - THOMAS,US539594,...,2006-ASO-6406-OE,<NA>,<NA>,"6650 Water Tank Town Road , Altoona, AL 35952,...",NaN,No,Active,Focus Asset,NaN,NaN


In [15]:
case_1_candidates['focus_asset_id'].nunique()

82

Thus, we only have 82 groupings that will fall under Case 1.

In [16]:
case_1_candidates.to_csv('case_1_candidates.csv')

In [17]:
case_1_candidates['focus_asset_id'].value_counts()

focus_asset_id
1146954    3
1147100    2
1147098    2
1147097    2
1147094    2
          ..
1142590    2
1142509    2
1142508    2
1142507    2
1147158    2
Name: count, Length: 82, dtype: Int64

---

Let's try to see how many groupings will fall under the <b>Case 2</b>.

<strong>Case 2: Both focus asset and associated asset/s have the same FCC ASR but have different FAA registry number </strong>

In [18]:
def find_case_2_candidates(df_with_differences):

    case_2_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']

            if pd.notnull(ref_fcc) and pd.notnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'] == ref_fcc) &
                    (paired_records_df['faa_study_number'] != ref_faa) &
                    (paired_records_df['fcc_asr_number'].notnull()) &
                    (paired_records_df['faa_study_number'].notnull())
                ]

                if not matching_paired_records.empty:
                    case_2_records.append(reference_record_df)
                    case_2_records.append(matching_paired_records)

    if case_2_records:
        case_2_candidates = pd.concat(case_2_records, ignore_index=True)
        return case_2_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_2_candidates = find_case_2_candidates(df_with_differences)
case_2_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1146953,Everest Infrastructure Partners,<NA>,Everest Infrastructure Partners,2025-02-18 16:21:07.530334,2025-02-18 16:21:07.465378,41.38513,-79.83374,101 Gurney Rd 2 of 2,US1030132,...,2000-AEA-1487-OE,<NA>,<NA>,"101 Gurney Rd , Franklin, PA 16323, UNITED STATES",NaN,No,Active,Focus Asset,NaN,NaN
1,1146953,Everest Infrastructure Partners,823488,ASR,2018-04-13 10:12:57.464868,2019-10-16 10:25:59.233356,41.38514,-79.83365,Franklin #2,01321,...,00-AEA-1487-OE,<NA>,<NA>,"End of The Gurney Road , Franklin, PA 16323, U...",1980-04-19,No,Active,Proximity Audit (7m),7.609403,1.4
2,1146954,Everest Infrastructure Partners,<NA>,Everest Infrastructure Partners,2025-02-18 16:21:09.455953,2025-02-18 16:21:09.388917,41.38513,-79.83374,101 Gurney Rd 1 of 2,US596342,...,2000-AEA-1487-OE,<NA>,<NA>,"101 Gurney Rd , Franklin, PA 16323, UNITED STATES",NaN,No,Active,Focus Asset,NaN,NaN
3,1146954,Everest Infrastructure Partners,823488,ASR,2018-04-13 10:12:57.464868,2019-10-16 10:25:59.233356,41.38514,-79.83365,Franklin #2,01321,...,00-AEA-1487-OE,<NA>,<NA>,"End of The Gurney Road , Franklin, PA 16323, U...",1980-04-19,No,Active,Proximity Audit (7m),7.609403,1.4


In [19]:
case_2_candidates['focus_asset_id'].nunique()

2

Thus, we only have 2 groupings that will fall under Case 2.

In [20]:
case_2_candidates.to_csv('case_2_candidates.csv')

---

Let's try to see how many groupings will fall under the <b>Case 3</b>.

<strong>Case 3: Both focus asset and associated asset/s have the same FCC ASR but the focus asset don’t have FAA while the associated asset/s have FAA </strong>


In [21]:
def find_case_3_candidates(df_with_differences):

    case_3_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']

            if pd.notnull(ref_fcc):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'] == ref_fcc) &
                    (paired_records_df['fcc_asr_number'].notnull()) &
                    (paired_records_df['faa_study_number'].notnull())
                ]

                if pd.isnull(ref_faa) and not matching_paired_records.empty:
                    case_3_records.append(reference_record_df)
                    case_3_records.append(matching_paired_records)

    if case_3_records:
        case_3_candidates = pd.concat(case_3_records, ignore_index=True)
        return case_3_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_3_candidates = find_case_3_candidates(df_with_differences)
case_3_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1147109,Everest Infrastructure Partners,<NA>,Everest Infrastructure Partners,2025-02-18 16:25:50.136163,2025-02-18 16:25:50.074986,33.97196,-84.50239,Blackjack Mountain 2,US701324,...,<NA>,<NA>,<NA>,"1480 Barnes Mill Road , Marietta, GA 30062, UN...",NaN,No,Dismantled,Focus Asset,NaN,NaN
1,1147109,Everest Infrastructure Partners,1136825,sgaither,2015-02-03 23:55:20.061725,2024-04-30 22:05:28.468952,33.97196,-84.50239,Blackjack Mountain 2,US701324,...,2002-ASO-293-OE,<NA>,Southeast,"1480 Barnes Mill Road , Marietta, GA 30062, UN...",1987-02-01,No,Dismantled,Manual Edit,0.0,0.0


In [22]:
case_3_candidates['focus_asset_id'].nunique()

1

Thus, we only have 1 grouping that will fall under Case 3.

In [23]:
case_3_candidates.to_csv('case_3_candidates.csv')

How about if vice-versa? (focus asset has the FAA while associated asset/s don't have)

In [24]:
def find_case_3_1_candidates(df_with_differences):

    case_3_1_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']

            if pd.notnull(ref_fcc) and pd.notnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'] == ref_fcc) &
                    (paired_records_df['fcc_asr_number'].notnull()) &
                    (paired_records_df['faa_study_number'].isnull())
                ]

                if not matching_paired_records.empty:
                    case_3_1_records.append(reference_record_df)
                    case_3_1_records.append(matching_paired_records)

    if case_3_1_records:
        case_3_1_candidates = pd.concat(case_3_1_records, ignore_index=True)
        return case_3_1_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_3_1_candidates = find_case_3_1_candidates(df_with_differences)
case_3_1_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1146945,Everest Infrastructure Partners,<NA>,Everest Infrastructure Partners,2025-02-18 16:20:56.672347,2025-02-18 16:20:56.605248,41.56311,-84.13123,1152 N Shoop Ave,US596309,...,1968-CLE-342-OE,<NA>,<NA>,"1152 N Shoop Ave , Wauseon, OH 43567, UNITED S...",NaN,No,Active,Focus Asset,NaN,NaN
1,1146945,Everest Infrastructure Partners,329248,dledbetter,2015-02-03 23:55:20.061725,2018-09-21 18:48:54.285743,41.56339,-84.13133,Waueson,01016,...,<NA>,<NA>,<NA>,"1302 North Shoop Avenue , Wauseon, OH 43567, U...",1965-01-01,No,Active,Proximity Audit (32m),32.197440,-0.1
2,1147003,Everest Infrastructure Partners,<NA>,Everest Infrastructure Partners,2025-02-18 16:22:01.924127,2025-02-18 16:22:01.857562,33.93656,-98.53389,3225 Maurine St,US596472,...,2015-ASW-7170-OE,<NA>,<NA>,"3225 Maurine St , WICHITA FALLS, TX 76306, UNI...",NaN,No,Active,Focus Asset,NaN,NaN
3,1147003,Everest Infrastructure Partners,327206,dledbetter,2015-02-03 23:55:20.061725,2018-09-21 19:48:20.430233,33.93653,-98.53391,Maurine Street,00583,...,<NA>,<NA>,<NA>,"3225 Maurine Street , Wichita Falls, TX 76305,...",1979-10-04,No,Active,Proximity Audit (3m),3.806865,-0.1


In [25]:
case_3_1_candidates['focus_asset_id'].nunique()

2

Thus, we only have 2 groupings that will fall under Case 3.1.

In [26]:
case_3_1_candidates.to_csv('case_3_1_candidates.csv')

---

Let's try to see how many groupings will fall under the <b>Case 4</b>.

<strong>Case 4: Focus site ID don’t have FCC ASR and FAA registry number but the associated asset/s have both numbers </strong>

In [27]:
def find_case_4_candidates(df_with_differences):

    case_4_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']

            if pd.isnull(ref_fcc) and pd.isnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'].notnull()) &
                    (paired_records_df['faa_study_number'].notnull())
                ]

                if not matching_paired_records.empty:
                    case_4_records.append(reference_record_df)
                    case_4_records.append(matching_paired_records)

    if case_4_records:
        case_4_candidates = pd.concat(case_4_records, ignore_index=True)
        return case_4_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_4_candidates = find_case_4_candidates(df_with_differences)
case_4_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,942761,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:58:33.921034,2023-01-13 12:41:13.613007,33.98899,-118.31900,CA12570,160568,...,<NA>,<NA>,<NA>,"SLAUSON AVE NL 42F E OF 3RD AVE EF , Los Angel...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
1,942761,Phoenix Tower International,524163,ASR,2016-02-18 21:35:39.242556,2016-02-18 16:20:23.876072,33.98894,-118.31922,2320 West Slauson Avenue,<NA>,...,2015-AWP-860-OE,<NA>,<NA>,"2320 West Slauson Avenue , Los Angeles, CA 900...",2015-11-09,No,Active,Proximity Audit (21m),21.070303,-10.9
2,944362,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 21:50:02.624769,2023-01-13 13:55:06.902417,33.96909,-118.25640,CA14222,162222,...,<NA>,<NA>,<NA>,"CENTRAL AVE EL 5F N OF 77TH PL N/S , Los Angel...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
3,944362,Phoenix Tower International,119688,dledbetter,2015-02-03 20:22:48.163448,2019-07-30 10:15:20.762731,33.96909,-118.25669,Griffin Property,CA40728-T,...,2018-AWP-10854-OE,<NA>,<NA>,", , CA 90001, UNITED STATES",2010-08-18,No,Active,Proximity Audit (26m),26.801288,-7.9
4,944536,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 21:55:30.147426,2023-01-13 11:20:03.272576,33.90079,-118.34300,CA14405,162405,...,<NA>,<NA>,<NA>,"PRAIRIE AVE EL 125F S OF ROSECRANS AVE N/S , H...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
25612,1161554,SBA,995053,ASR,2020-06-01 08:55:02.975387,2020-06-01 14:08:59.76073,34.11269,-88.02347,Firetower Rd,<NA>,...,2019-ASO-1238-OE,<NA>,<NA>,"Firetower Rd , Hamilton, AL 35581, UNITED STATES",2021-04-14,No,Active,Proximity Audit (41m),41.589267,-4.6
25613,1161555,SBA,<NA>,SBA,2025-02-20 05:42:09.896726,2025-02-20 05:42:09.835937,33.95821,-87.83784,Winfield 2 AL,AL04759-A,...,<NA>,<NA>,<NA>,"2501 State Highway 253 , Winfield, AL 35594, U...",NaN,No,Active,Focus Asset,NaN,NaN
25614,1161555,SBA,118025,SBA,2015-02-03 20:22:46.439621,2017-11-08 10:39:31.620294,33.95821,-87.83785,"Winfield 2, AL",AL04759-A,...,2025-ASO-6242-OE,<NA>,<NA>,", Winfield, AL 35594, UNITED STATES",2000-06-15,No,Active,Proximity Audit (0m),0.924300,0.0
25615,1161556,SBA,<NA>,SBA,2025-02-20 05:42:15.236684,2025-02-20 05:42:15.173002,34.58570,-86.45280,Hamptons Cove,AL04832-B,...,<NA>,<NA>,<NA>,"238 Wilson Mann Road , Owens Cross Roads, AL 3...",NaN,No,Active,Focus Asset,NaN,NaN


In [28]:
case_4_candidates['focus_asset_id'].nunique()

12553

We have 12,553 groupings falling under Case 4. Now, let's try to at least add more restrictions to the logic. Let's add Operator as one of the matching condition.

In [29]:
def find_case_4_1_candidates(df_with_differences):

    case_4_1_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']
            ref_operator = reference_record['operator_name']

            if pd.isnull(ref_fcc) and pd.isnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'].notnull()) &
                    (paired_records_df['faa_study_number'].notnull()) &
                    (paired_records_df['operator_name'] == ref_operator)
                ]

                if not matching_paired_records.empty:
                    case_4_1_records.append(reference_record_df)
                    case_4_1_records.append(matching_paired_records)

    if case_4_1_records:
        case_4_1_candidates = pd.concat(case_4_1_records, ignore_index=True)
        return case_4_1_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_4_1_candidates = find_case_4_1_candidates(df_with_differences)
case_4_1_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1055409,Capital Telecom,<NA>,Capital Telecom,2022-10-26 17:50:57.477363,2022-10-26 13:50:57.315393,41.77024,-88.20001,CT1800,<NA>,...,<NA>,<NA>,<NA>,"1847 West Jefferson Avenue , Naperville, IL 60...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
1,1055409,Capital Telecom,828023,adarr,2016-04-08 19:52:19.823551,2018-07-20 20:43:42.836932,41.77024,-88.20001,1847 West Jefferson Avenue,CT1800,...,2015-AGL-17599-OE,<NA>,<NA>,"1847 West Jefferson Avenue , Naperville, IL 60...",2016-11-21,No,Active,Manual Edit,0.000000,NaN
2,1055421,Capital Telecom,<NA>,Capital Telecom,2022-10-26 17:51:41.467623,2022-10-26 19:49:40.512964,40.15238,-80.28743,Highway 70 & Ridgewood Drive,CT1842,...,<NA>,<NA>,<NA>,"Highway 70 & Ridgewood Drive , Washington, PA ...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
3,1055421,Capital Telecom,820926,adarr,2016-04-08 19:52:19.823551,2018-07-17 19:57:17.642071,40.15256,-80.28764,Highway 70 & Ridgewood Drive,CT1841,...,2016-AEA-5598-OE,<NA>,<NA>,"200 Canyon Drive , Washington, PA 15301, UNITE...",2017-10-23,No,Active,Proximity Audit (26m),26.825763,NaN
4,1055438,Capital Telecom,<NA>,Capital Telecom,2022-10-26 17:52:35.270568,2022-10-26 13:52:35.158531,40.77773,-82.51294,CT1844,<NA>,...,<NA>,<NA>,<NA>,"41 Longview Avenue East , Mansfield, OH 44903,...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
23053,1161554,SBA,118022,SBA,2015-02-03 20:22:46.433772,2017-11-08 13:03:54.697111,34.11235,-88.02329,West Hamilton,AL04756-A,...,2024-ASO-19958-OE,<NA>,<NA>,", Hamilton, AL 35570-9721, UNITED STATES",2000-06-24,No,Active,Proximity Audit (0m),0.922630,-3.1
23054,1161555,SBA,<NA>,SBA,2025-02-20 05:42:09.896726,2025-02-20 05:42:09.835937,33.95821,-87.83784,Winfield 2 AL,AL04759-A,...,<NA>,<NA>,<NA>,"2501 State Highway 253 , Winfield, AL 35594, U...",NaN,No,Active,Focus Asset,NaN,NaN
23055,1161555,SBA,118025,SBA,2015-02-03 20:22:46.439621,2017-11-08 10:39:31.620294,33.95821,-87.83785,"Winfield 2, AL",AL04759-A,...,2025-ASO-6242-OE,<NA>,<NA>,", Winfield, AL 35594, UNITED STATES",2000-06-15,No,Active,Proximity Audit (0m),0.924300,0.0
23056,1161556,SBA,<NA>,SBA,2025-02-20 05:42:15.236684,2025-02-20 05:42:15.173002,34.58570,-86.45280,Hamptons Cove,AL04832-B,...,<NA>,<NA>,<NA>,"238 Wilson Mann Road , Owens Cross Roads, AL 3...",NaN,No,Active,Focus Asset,NaN,NaN


In [30]:
case_4_1_candidates['focus_asset_id'].nunique()

11474

From 12,553, now we're down to 11,474 groupings. Let's try to add Operator site ID as one of the matching condition.

In [31]:
def find_case_4_2_candidates(df_with_differences):

    case_4_2_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']
            ref_operator = reference_record['operator_name']
            ref_site_id = reference_record['operator_site_id']

            if pd.isnull(ref_fcc) and pd.isnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'].notnull()) &
                    (paired_records_df['faa_study_number'].notnull()) &
                    (paired_records_df['operator_name'] == ref_operator) &
                    (paired_records_df['operator_site_id'] == ref_site_id)
                ]

                if not matching_paired_records.empty:
                    case_4_2_records.append(reference_record_df)
                    case_4_2_records.append(matching_paired_records)

    if case_4_2_records:
        case_4_2_candidates = pd.concat(case_4_2_records, ignore_index=True)
        return case_4_2_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_4_2_candidates = find_case_4_2_candidates(df_with_differences)
case_4_2_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1055447,Capital Telecom,<NA>,Capital Telecom,2022-10-26 17:52:49.989083,2022-10-26 19:50:27.864634,40.85599,-79.48862,323 Cowanshonnock Road,CT1870,...,<NA>,<NA>,<NA>,"323 Cowanshonnock Road , Kittanning, PA 16201,...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
1,1055447,Capital Telecom,808691,dledbetter,2016-07-18 21:11:07.103328,2018-07-23 18:12:22.928009,40.85600,-79.48861,221 Cowanshannock Rd.,CT1870,...,2017-AEA-1996-OE,<NA>,<NA>,"323 Cowanshonnock Road , Kittanning, PA 16201,...",NaN,No,Active,Proximity Audit (1m),1.394342,NaN
2,1055511,Capital Telecom,<NA>,Capital Telecom,2022-10-26 17:55:42.182028,2022-10-26 19:52:51.475428,38.23051,-85.76771,901 West Magnolia Avenue,CT1788,...,<NA>,<NA>,<NA>,"901 West Magnolia Avenue , Louisville, KY 4020...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
3,1055511,Capital Telecom,642619,dledbetter,2016-11-29 17:31:01.94261,2018-07-20 22:17:46.886541,38.23050,-85.76772,901 West Magnolia Avenue,CT1788,...,2016-ASO-7894-OE,<NA>,<NA>,"901 West Magnolia Avenue , Louisville, KY 4020...",2017-02-27,No,Active,Proximity Audit (1m),1.413767,NaN
4,1058685,Capital Telecom,<NA>,Capital Telecom,2022-10-26 23:43:01.522002,2022-10-26 19:43:01.458228,38.11985,-85.48786,15110 Brush Run Road,CT1732,...,<NA>,<NA>,<NA>,"15110 Brush Run Road , Jeffersontown, KY 40299...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
21627,1161554,SBA,118022,SBA,2015-02-03 20:22:46.433772,2017-11-08 13:03:54.697111,34.11235,-88.02329,West Hamilton,AL04756-A,...,2024-ASO-19958-OE,<NA>,<NA>,", Hamilton, AL 35570-9721, UNITED STATES",2000-06-24,No,Active,Proximity Audit (0m),0.922630,-3.1
21628,1161555,SBA,<NA>,SBA,2025-02-20 05:42:09.896726,2025-02-20 05:42:09.835937,33.95821,-87.83784,Winfield 2 AL,AL04759-A,...,<NA>,<NA>,<NA>,"2501 State Highway 253 , Winfield, AL 35594, U...",NaN,No,Active,Focus Asset,NaN,NaN
21629,1161555,SBA,118025,SBA,2015-02-03 20:22:46.439621,2017-11-08 10:39:31.620294,33.95821,-87.83785,"Winfield 2, AL",AL04759-A,...,2025-ASO-6242-OE,<NA>,<NA>,", Winfield, AL 35594, UNITED STATES",2000-06-15,No,Active,Proximity Audit (0m),0.924300,0.0
21630,1161556,SBA,<NA>,SBA,2025-02-20 05:42:15.236684,2025-02-20 05:42:15.173002,34.58570,-86.45280,Hamptons Cove,AL04832-B,...,<NA>,<NA>,<NA>,"238 Wilson Mann Road , Owens Cross Roads, AL 3...",NaN,No,Active,Focus Asset,NaN,NaN


In [32]:
case_4_2_candidates['focus_asset_id'].nunique()

10794

Now we're down to 10,794 groupings. Let's add tower status now and see how many groupings will be left.

In [33]:
def find_case_4_3_candidates(df_with_differences):

    case_4_3_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']
            ref_operator = reference_record['operator_name']
            ref_site_id = reference_record['operator_site_id']
            ref_status = reference_record['asset_status']

            if pd.isnull(ref_fcc) and pd.isnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'].notnull()) &
                    (paired_records_df['faa_study_number'].notnull()) &
                    (paired_records_df['operator_name'] == ref_operator) &
                    (paired_records_df['operator_site_id'] == ref_site_id) &
                    (paired_records_df['asset_status'] == ref_status)
                ]

                if not matching_paired_records.empty:
                    case_4_3_records.append(reference_record_df)
                    case_4_3_records.append(matching_paired_records)

    if case_4_3_records:
        case_4_3_candidates = pd.concat(case_4_3_records, ignore_index=True)
        return case_4_3_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_4_3_candidates = find_case_4_3_candidates(df_with_differences)
case_4_3_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1147159,SBA,<NA>,SBA,2025-02-19 20:47:14.736408,2025-02-19 20:47:14.612784,61.11381,-149.86151,Canine Castle,AK12386-A,...,<NA>,<NA>,<NA>,"11801 Old Seward Hwy , Anchorage, AK 99515, UN...",NaN,No,Active,Focus Asset,NaN,NaN
1,1147159,SBA,117947,dledbetter,2015-02-03 20:22:46.31056,2019-07-30 10:15:20.762731,61.11382,-149.86151,Canine Castle,AK12386-A,...,2010-AAL-202-OE,<NA>,<NA>,", Anchorage, AK 99515, UNITED STATES",2009-05-06,No,Active,Proximity Audit (1m),1.114310,0.0
2,1147160,SBA,<NA>,SBA,2025-02-19 20:47:16.835983,2025-02-19 20:47:16.766783,61.16028,-149.85808,Brayton Pax Soto,AK12387-A,...,<NA>,<NA>,<NA>,"6646 Homer Drive , Anchorage, AK 99518, UNITED...",NaN,No,Active,Focus Asset,NaN,NaN
3,1147160,SBA,117948,dledbetter,2015-02-03 20:22:46.346536,2019-07-30 10:15:20.762731,61.16028,-149.85809,Brayton Pax Soto,AK12387-A,...,2010-AAL-203-OE,<NA>,<NA>,", Anchorage, AK 99518, UNITED STATES",2009-05-06,No,Active,Proximity Audit (0m),0.538346,0.0
4,1147161,SBA,<NA>,SBA,2025-02-19 20:47:18.928215,2025-02-19 20:47:18.862005,61.13868,-149.84196,Tikigaq,AK12388-A,...,<NA>,<NA>,<NA>,"2121 Abbott Road , Anchorage, AK 99507, UNITED...",NaN,No,Active,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
21501,1161554,SBA,118022,SBA,2015-02-03 20:22:46.433772,2017-11-08 13:03:54.697111,34.11235,-88.02329,West Hamilton,AL04756-A,...,2024-ASO-19958-OE,<NA>,<NA>,", Hamilton, AL 35570-9721, UNITED STATES",2000-06-24,No,Active,Proximity Audit (0m),0.922630,-3.1
21502,1161555,SBA,<NA>,SBA,2025-02-20 05:42:09.896726,2025-02-20 05:42:09.835937,33.95821,-87.83784,Winfield 2 AL,AL04759-A,...,<NA>,<NA>,<NA>,"2501 State Highway 253 , Winfield, AL 35594, U...",NaN,No,Active,Focus Asset,NaN,NaN
21503,1161555,SBA,118025,SBA,2015-02-03 20:22:46.439621,2017-11-08 10:39:31.620294,33.95821,-87.83785,"Winfield 2, AL",AL04759-A,...,2025-ASO-6242-OE,<NA>,<NA>,", Winfield, AL 35594, UNITED STATES",2000-06-15,No,Active,Proximity Audit (0m),0.924300,0.0
21504,1161556,SBA,<NA>,SBA,2025-02-20 05:42:15.236684,2025-02-20 05:42:15.173002,34.58570,-86.45280,Hamptons Cove,AL04832-B,...,<NA>,<NA>,<NA>,"238 Wilson Mann Road , Owens Cross Roads, AL 3...",NaN,No,Active,Focus Asset,NaN,NaN


In [34]:
case_4_3_candidates['focus_asset_id'].nunique()

10731

Now we're down to 10,731 groupings. Let's add manager now and see how many groupings will be left.

In [35]:
def find_case_4_4_candidates(df_with_differences):

    case_4_4_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']
            ref_operator = reference_record['operator_name']
            ref_site_id = reference_record['operator_site_id']
            ref_status = reference_record['asset_status']
            ref_type = reference_record['type']

            if pd.isnull(ref_fcc) and pd.isnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'].notnull()) &
                    (paired_records_df['faa_study_number'].notnull()) &
                    (paired_records_df['operator_name'] == ref_operator) &
                    (paired_records_df['operator_site_id'] == ref_site_id) &
                    (paired_records_df['asset_status'] == ref_status) &
                    (paired_records_df['type'] == ref_type)
                ]

                if not matching_paired_records.empty:
                    case_4_4_records.append(reference_record_df)
                    case_4_4_records.append(matching_paired_records)

    if case_4_4_records:
        case_4_4_candidates = pd.concat(case_4_4_records, ignore_index=True)
        return case_4_4_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_4_4_candidates = find_case_4_4_candidates(df_with_differences)
case_4_4_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1147159,SBA,<NA>,SBA,2025-02-19 20:47:14.736408,2025-02-19 20:47:14.612784,61.11381,-149.86151,Canine Castle,AK12386-A,...,<NA>,<NA>,<NA>,"11801 Old Seward Hwy , Anchorage, AK 99515, UN...",NaN,No,Active,Focus Asset,NaN,NaN
1,1147159,SBA,117947,dledbetter,2015-02-03 20:22:46.31056,2019-07-30 10:15:20.762731,61.11382,-149.86151,Canine Castle,AK12386-A,...,2010-AAL-202-OE,<NA>,<NA>,", Anchorage, AK 99515, UNITED STATES",2009-05-06,No,Active,Proximity Audit (1m),1.114310,0.0
2,1147160,SBA,<NA>,SBA,2025-02-19 20:47:16.835983,2025-02-19 20:47:16.766783,61.16028,-149.85808,Brayton Pax Soto,AK12387-A,...,<NA>,<NA>,<NA>,"6646 Homer Drive , Anchorage, AK 99518, UNITED...",NaN,No,Active,Focus Asset,NaN,NaN
3,1147160,SBA,117948,dledbetter,2015-02-03 20:22:46.346536,2019-07-30 10:15:20.762731,61.16028,-149.85809,Brayton Pax Soto,AK12387-A,...,2010-AAL-203-OE,<NA>,<NA>,", Anchorage, AK 99518, UNITED STATES",2009-05-06,No,Active,Proximity Audit (0m),0.538346,0.0
4,1147161,SBA,<NA>,SBA,2025-02-19 20:47:18.928215,2025-02-19 20:47:18.862005,61.13868,-149.84196,Tikigaq,AK12388-A,...,<NA>,<NA>,<NA>,"2121 Abbott Road , Anchorage, AK 99507, UNITED...",NaN,No,Active,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
20677,1161554,SBA,118022,SBA,2015-02-03 20:22:46.433772,2017-11-08 13:03:54.697111,34.11235,-88.02329,West Hamilton,AL04756-A,...,2024-ASO-19958-OE,<NA>,<NA>,", Hamilton, AL 35570-9721, UNITED STATES",2000-06-24,No,Active,Proximity Audit (0m),0.922630,-3.1
20678,1161555,SBA,<NA>,SBA,2025-02-20 05:42:09.896726,2025-02-20 05:42:09.835937,33.95821,-87.83784,Winfield 2 AL,AL04759-A,...,<NA>,<NA>,<NA>,"2501 State Highway 253 , Winfield, AL 35594, U...",NaN,No,Active,Focus Asset,NaN,NaN
20679,1161555,SBA,118025,SBA,2015-02-03 20:22:46.439621,2017-11-08 10:39:31.620294,33.95821,-87.83785,"Winfield 2, AL",AL04759-A,...,2025-ASO-6242-OE,<NA>,<NA>,", Winfield, AL 35594, UNITED STATES",2000-06-15,No,Active,Proximity Audit (0m),0.924300,0.0
20680,1161556,SBA,<NA>,SBA,2025-02-20 05:42:15.236684,2025-02-20 05:42:15.173002,34.58570,-86.45280,Hamptons Cove,AL04832-B,...,<NA>,<NA>,<NA>,"238 Wilson Mann Road , Owens Cross Roads, AL 3...",NaN,No,Active,Focus Asset,NaN,NaN


In [36]:
case_4_4_candidates['focus_asset_id'].nunique()

10327

Now we're down to 10,327 groupings. Let's add tower name now and see how many groupings will be left.

In [37]:
def find_case_4_5_candidates(df_with_differences):

    case_4_5_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']
            ref_operator = reference_record['operator_name']
            ref_site_id = reference_record['operator_site_id']
            ref_status = reference_record['asset_status']
            ref_type = reference_record['type']
            ref_tower_name = reference_record['name']

            if pd.isnull(ref_fcc) and pd.isnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'].notnull()) &
                    (paired_records_df['faa_study_number'].notnull()) &
                    (paired_records_df['operator_name'] == ref_operator) &
                    (paired_records_df['operator_site_id'] == ref_site_id) &
                    (paired_records_df['asset_status'] == ref_status) &
                    (paired_records_df['type'] == ref_type) &
                    (paired_records_df['name'] == ref_tower_name)
                ]

                if not matching_paired_records.empty:
                    case_4_5_records.append(reference_record_df)
                    case_4_5_records.append(matching_paired_records)

    if case_4_5_records:
        case_4_5_candidates = pd.concat(case_4_5_records, ignore_index=True)
        return case_4_5_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_4_5_candidates = find_case_4_5_candidates(df_with_differences)
case_4_5_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1147159,SBA,<NA>,SBA,2025-02-19 20:47:14.736408,2025-02-19 20:47:14.612784,61.11381,-149.86151,Canine Castle,AK12386-A,...,<NA>,<NA>,<NA>,"11801 Old Seward Hwy , Anchorage, AK 99515, UN...",NaN,No,Active,Focus Asset,NaN,NaN
1,1147159,SBA,117947,dledbetter,2015-02-03 20:22:46.31056,2019-07-30 10:15:20.762731,61.11382,-149.86151,Canine Castle,AK12386-A,...,2010-AAL-202-OE,<NA>,<NA>,", Anchorage, AK 99515, UNITED STATES",2009-05-06,No,Active,Proximity Audit (1m),1.114310,0.0
2,1147160,SBA,<NA>,SBA,2025-02-19 20:47:16.835983,2025-02-19 20:47:16.766783,61.16028,-149.85808,Brayton Pax Soto,AK12387-A,...,<NA>,<NA>,<NA>,"6646 Homer Drive , Anchorage, AK 99518, UNITED...",NaN,No,Active,Focus Asset,NaN,NaN
3,1147160,SBA,117948,dledbetter,2015-02-03 20:22:46.346536,2019-07-30 10:15:20.762731,61.16028,-149.85809,Brayton Pax Soto,AK12387-A,...,2010-AAL-203-OE,<NA>,<NA>,", Anchorage, AK 99518, UNITED STATES",2009-05-06,No,Active,Proximity Audit (0m),0.538346,0.0
4,1147161,SBA,<NA>,SBA,2025-02-19 20:47:18.928215,2025-02-19 20:47:18.862005,61.13868,-149.84196,Tikigaq,AK12388-A,...,<NA>,<NA>,<NA>,"2121 Abbott Road , Anchorage, AK 99507, UNITED...",NaN,No,Active,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15689,1161552,SBA,118020,SBA,2015-02-03 20:22:46.431704,2017-11-08 13:04:11.713717,33.75752,-85.78392,Hamiltons Lake,AL04110-B,...,2022-ASO-34760-OE,<NA>,<NA>,", Anniston, AL 36206-7568, UNITED STATES",2000-08-03,No,Active,Proximity Audit (1m),1.109180,-0.1
15690,1161554,SBA,<NA>,SBA,2025-02-20 05:42:05.608936,2025-02-20 05:42:05.545048,34.11235,-88.02328,West Hamilton,AL04756-A,...,<NA>,<NA>,<NA>,"596 Fire Tower Road , Hamilton, AL 35570-9721,...",NaN,No,Active,Focus Asset,NaN,NaN
15691,1161554,SBA,118022,SBA,2015-02-03 20:22:46.433772,2017-11-08 13:03:54.697111,34.11235,-88.02329,West Hamilton,AL04756-A,...,2024-ASO-19958-OE,<NA>,<NA>,", Hamilton, AL 35570-9721, UNITED STATES",2000-06-24,No,Active,Proximity Audit (0m),0.922630,-3.1
15692,1161556,SBA,<NA>,SBA,2025-02-20 05:42:15.236684,2025-02-20 05:42:15.173002,34.58570,-86.45280,Hamptons Cove,AL04832-B,...,<NA>,<NA>,<NA>,"238 Wilson Mann Road , Owens Cross Roads, AL 3...",NaN,No,Active,Focus Asset,NaN,NaN


In [38]:
case_4_5_candidates['focus_asset_id'].nunique()

7841

In [39]:
case_4_1_candidates.to_csv('case_4_1_candidates.csv')
case_4_2_candidates.to_csv('case_4_2_candidates.csv')
case_4_3_candidates.to_csv('case_4_3_candidates.csv')
case_4_4_candidates.to_csv('case_4_4_candidates.csv')
case_4_5_candidates.to_csv('case_4_5_candidates.csv')

Now we're down to 7,841 groupings

---

Let's try to see how many groupings will fall under the <b>Case 5</b>.

<strong>Case 5: Both focus asset and associated asset/s don’t have FCC ASR and FAA registry number</strong>

Let's try to mimic what we did on Case 4 where we added parameters in our matching condition one at a time to see how each parameter helped.

Let's do matching first on FCC ASR and Registry number to be both NULL between reference record (focus asset) and paired record/s (associated asset/s)

In [40]:
def find_case_5_candidates(df_with_differences):

    case_5_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']

            if pd.isnull(ref_fcc) and pd.isnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'].isnull()) &
                    (paired_records_df['faa_study_number'].isnull())
                ]

                if not matching_paired_records.empty:
                    case_5_records.append(reference_record_df)
                    case_5_records.append(matching_paired_records)

    if case_5_records:
        case_5_candidates = pd.concat(case_5_records, ignore_index=True)
        return case_5_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_5_candidates = find_case_5_candidates(df_with_differences)
case_5_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,942614,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:53:42.284115,2023-01-13 15:25:12.493704,33.94619,-118.28200,CA12409,160407,...,<NA>,<NA>,<NA>,"FIGUEROA ST WL 75F S OF 99TH ST S/N , August F...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
1,942614,Phoenix Tower International,943927,Phoenix Tower International,2019-08-22 21:36:14.93064,2023-06-26 12:05:50.800659,33.94574,-118.28200,MAIN ST EL 185F N OF 110TH ST SF,161773,...,<NA>,<NA>,<NA>,"MAIN ST EL 185F N OF 110TH ST SF , , CA 90003,...",NaN,No,Unconfirmed,Proximity Audit (48m),49.914635,0.0
2,942664,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:55:21.34445,2023-01-13 11:49:35.586868,34.03929,-118.19900,CA12465,160463,...,<NA>,<NA>,<NA>,"1ST ST SL 2F E OF DACOTAH ST E/W , Los Angeles...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
3,942664,Phoenix Tower International,942203,Phoenix Tower International,2019-08-22 20:40:05.047291,2023-01-13 13:04:52.922687,34.03910,-118.19900,CA11987,159985,...,<NA>,<NA>,<NA>,"1ST ST SL 53F W OF FRESNO ST WF , Los Angeles,...",NaN,No,Unconfirmed,Proximity Audit (21m),21.075388,0.0
4,942761,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:58:33.921034,2023-01-13 12:41:13.613007,33.98899,-118.31900,CA12570,160568,...,<NA>,<NA>,<NA>,"SLAUSON AVE NL 42F E OF 3RD AVE EF , Los Angel...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
38536,1161507,SBA,118581,SBA,2015-02-03 20:22:47.027416,2017-11-08 10:41:30.37165,33.03370,-85.77584,"New Site 2, AL",AL40493-T,...,<NA>,<NA>,<NA>,", New Site, AL 36256, UNITED STATES",NaN,No,Active,Proximity Audit (25m),25.222771,0.0
38537,1161508,SBA,<NA>,SBA,2025-02-20 05:40:16.728972,2025-02-20 05:40:16.66007,33.07055,-86.04802,Goodwater 2 AL,AL40494-T,...,<NA>,<NA>,<NA>,"22881 AL Hwy 9 , Goodwater, AL 35072, UNITED S...",NaN,No,Active,Focus Asset,NaN,NaN
38538,1161508,SBA,118582,SBA,2015-02-03 20:22:47.028351,2017-11-08 10:37:58.999645,33.07056,-86.04802,"Goodwater 2, AL",AL40494-T,...,<NA>,<NA>,<NA>,", Goodwater, AL 35072, UNITED STATES",NaN,No,Active,Proximity Audit (1m),1.109057,0.0
38539,1161551,SBA,<NA>,SBA,2025-02-20 05:41:59.208766,2025-02-20 05:41:59.143873,33.63757,-85.80192,Coleman Road,AL04109-B,...,<NA>,<NA>,<NA>,"2510 Coleman Road , Anniston, AL 36207, UNITED...",NaN,No,Active,Focus Asset,NaN,NaN


In [41]:
case_5_candidates['focus_asset_id'].nunique()

16114

We got 16,114 groupings for this condition where both focus asset and associated asset/s have NULL FCC ASR and FAA registry number. Now, let's add operator name as one of the matching condition.

In [42]:
def find_case_5_1_candidates(df_with_differences):

    case_5_1_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']
            ref_operator = reference_record['operator_name']

            if pd.isnull(ref_fcc) and pd.isnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'].isnull()) &
                    (paired_records_df['faa_study_number'].isnull()) &
                    (paired_records_df['operator_name'] == ref_operator)
                ]

                if not matching_paired_records.empty:
                    case_5_1_records.append(reference_record_df)
                    case_5_1_records.append(matching_paired_records)

    if case_5_1_records:
        case_5_1_candidates = pd.concat(case_5_1_records, ignore_index=True)
        return case_5_1_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_5_1_candidates = find_case_5_1_candidates(df_with_differences)
case_5_1_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,942614,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:53:42.284115,2023-01-13 15:25:12.493704,33.94619,-118.28200,CA12409,160407,...,<NA>,<NA>,<NA>,"FIGUEROA ST WL 75F S OF 99TH ST S/N , August F...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
1,942614,Phoenix Tower International,943927,Phoenix Tower International,2019-08-22 21:36:14.93064,2023-06-26 12:05:50.800659,33.94574,-118.28200,MAIN ST EL 185F N OF 110TH ST SF,161773,...,<NA>,<NA>,<NA>,"MAIN ST EL 185F N OF 110TH ST SF , , CA 90003,...",NaN,No,Unconfirmed,Proximity Audit (48m),49.914635,0.0
2,942664,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:55:21.34445,2023-01-13 11:49:35.586868,34.03929,-118.19900,CA12465,160463,...,<NA>,<NA>,<NA>,"1ST ST SL 2F E OF DACOTAH ST E/W , Los Angeles...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
3,942664,Phoenix Tower International,942203,Phoenix Tower International,2019-08-22 20:40:05.047291,2023-01-13 13:04:52.922687,34.03910,-118.19900,CA11987,159985,...,<NA>,<NA>,<NA>,"1ST ST SL 53F W OF FRESNO ST WF , Los Angeles,...",NaN,No,Unconfirmed,Proximity Audit (21m),21.075388,0.0
4,942761,Phoenix Tower International,<NA>,Phoenix Tower International,2019-08-22 20:58:33.921034,2023-01-13 12:41:13.613007,33.98899,-118.31900,CA12570,160568,...,<NA>,<NA>,<NA>,"SLAUSON AVE NL 42F E OF 3RD AVE EF , Los Angel...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
21894,1161490,SBA,118525,SBA,2015-02-03 20:22:46.964012,2017-11-08 10:40:17.274368,34.83366,-86.64344,Burwell Springs,AL40003-A,...,<NA>,<NA>,<NA>,", Huntsville, AL 35773, UNITED STATES",NaN,No,Active,Proximity Audit (0m),0.914725,0.9
21895,1161507,SBA,<NA>,SBA,2025-02-20 05:40:14.529083,2025-02-20 05:40:14.465301,33.03370,-85.77611,New Site 2 AL,AL40493-T,...,<NA>,<NA>,<NA>,"177 Tower Street , New Site, AL 36256, UNITED ...",NaN,No,Active,Focus Asset,NaN,NaN
21896,1161507,SBA,118581,SBA,2015-02-03 20:22:47.027416,2017-11-08 10:41:30.37165,33.03370,-85.77584,"New Site 2, AL",AL40493-T,...,<NA>,<NA>,<NA>,", New Site, AL 36256, UNITED STATES",NaN,No,Active,Proximity Audit (25m),25.222771,0.0
21897,1161508,SBA,<NA>,SBA,2025-02-20 05:40:16.728972,2025-02-20 05:40:16.66007,33.07055,-86.04802,Goodwater 2 AL,AL40494-T,...,<NA>,<NA>,<NA>,"22881 AL Hwy 9 , Goodwater, AL 35072, UNITED S...",NaN,No,Active,Focus Asset,NaN,NaN


In [43]:
case_5_1_candidates['focus_asset_id'].nunique()

9205

We got 9,205 groupings for this condition. Now let's add Operator Site ID as one of the matching condition.

In [44]:
def find_case_5_2_candidates(df_with_differences):

    case_5_2_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']
            ref_operator = reference_record['operator_name']
            ref_site_id = reference_record['operator_site_id']

            if pd.isnull(ref_fcc) and pd.isnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'].isnull()) &
                    (paired_records_df['faa_study_number'].isnull()) &
                    (paired_records_df['operator_name'] == ref_operator) &
                    (paired_records_df['operator_site_id'] == ref_site_id)
                ]

                if not matching_paired_records.empty:
                    case_5_2_records.append(reference_record_df)
                    case_5_2_records.append(matching_paired_records)

    if case_5_2_records:
        case_5_2_candidates = pd.concat(case_5_2_records, ignore_index=True)
        return case_5_2_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_5_2_candidates = find_case_5_2_candidates(df_with_differences)
case_5_2_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1053591,Phoenix Tower International,<NA>,Phoenix Tower International,2022-10-26 15:53:01.59936,2023-06-26 12:05:50.800659,27.95261,-82.36364,8326 East Adamo Drive,AMBPRO214,...,<NA>,<NA>,<NA>,"8326 East Adamo Drive , Tampa, FL 33619, UNITE...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
1,1053591,Phoenix Tower International,192180,Phoenix Tower International,2015-03-02 23:52:33.697922,2023-06-26 12:05:50.800659,27.95261,-82.36364,8326 East Adamo Drive,AMBPRO214,...,<NA>,<NA>,<NA>,"8326 East Adamo Drive , Tampa, FL 33619, UNITE...",NaN,No,Unconfirmed,Proximity Audit (0m),0.000000,0.0
2,1053602,Phoenix Tower International,<NA>,Phoenix Tower International,2022-10-26 15:53:59.431417,2023-06-26 12:05:50.800659,35.19402,-80.88992,4123 Barringer Drive,AMBPRO387,...,<NA>,<NA>,<NA>,"4123 Barringer Drive , Charlotte, NC 28217, UN...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
3,1053602,Phoenix Tower International,192250,Phoenix Tower International,2015-03-02 23:52:33.804113,2023-06-26 12:05:50.800659,35.19402,-80.88992,4123 Barringer Drive,AMBPRO387,...,<NA>,<NA>,<NA>,"4123 Barringer Drive , Charlotte, NC 28217, UN...",NaN,No,Unconfirmed,Proximity Audit (1m),0.000000,0.0
4,1054266,Phoenix Tower International,<NA>,Phoenix Tower International,2022-10-26 16:35:12.716918,2023-06-26 12:05:50.800659,40.40552,-80.55032,1360 Cove Road,DDR604,...,<NA>,<NA>,<NA>,"1360 Cove Road , Weirton, WV 26062, UNITED STATES",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9628,1161490,SBA,118525,SBA,2015-02-03 20:22:46.964012,2017-11-08 10:40:17.274368,34.83366,-86.64344,Burwell Springs,AL40003-A,...,<NA>,<NA>,<NA>,", Huntsville, AL 35773, UNITED STATES",NaN,No,Active,Proximity Audit (0m),0.914725,0.9
9629,1161507,SBA,<NA>,SBA,2025-02-20 05:40:14.529083,2025-02-20 05:40:14.465301,33.03370,-85.77611,New Site 2 AL,AL40493-T,...,<NA>,<NA>,<NA>,"177 Tower Street , New Site, AL 36256, UNITED ...",NaN,No,Active,Focus Asset,NaN,NaN
9630,1161507,SBA,118581,SBA,2015-02-03 20:22:47.027416,2017-11-08 10:41:30.37165,33.03370,-85.77584,"New Site 2, AL",AL40493-T,...,<NA>,<NA>,<NA>,", New Site, AL 36256, UNITED STATES",NaN,No,Active,Proximity Audit (25m),25.222771,0.0
9631,1161508,SBA,<NA>,SBA,2025-02-20 05:40:16.728972,2025-02-20 05:40:16.66007,33.07055,-86.04802,Goodwater 2 AL,AL40494-T,...,<NA>,<NA>,<NA>,"22881 AL Hwy 9 , Goodwater, AL 35072, UNITED S...",NaN,No,Active,Focus Asset,NaN,NaN


In [45]:
case_5_2_candidates['focus_asset_id'].nunique()

4727

We got 4,727 groupings for this condition. Now let's add Asset Status as one of the matching condition.

In [46]:
def find_case_5_3_candidates(df_with_differences):

    case_5_3_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']
            ref_operator = reference_record['operator_name']
            ref_site_id = reference_record['operator_site_id']
            ref_status = reference_record['asset_status']

            if pd.isnull(ref_fcc) and pd.isnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'].isnull()) &
                    (paired_records_df['faa_study_number'].isnull()) &
                    (paired_records_df['operator_name'] == ref_operator) &
                    (paired_records_df['operator_site_id'] == ref_site_id) &
                    (paired_records_df['asset_status'] == ref_status)
                ]

                if not matching_paired_records.empty:
                    case_5_3_records.append(reference_record_df)
                    case_5_3_records.append(matching_paired_records)

    if case_5_3_records:
        case_5_3_candidates = pd.concat(case_5_3_records, ignore_index=True)
        return case_5_3_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_5_3_candidates = find_case_5_3_candidates(df_with_differences)
case_5_3_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1053591,Phoenix Tower International,<NA>,Phoenix Tower International,2022-10-26 15:53:01.59936,2023-06-26 12:05:50.800659,27.95261,-82.36364,8326 East Adamo Drive,AMBPRO214,...,<NA>,<NA>,<NA>,"8326 East Adamo Drive , Tampa, FL 33619, UNITE...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
1,1053591,Phoenix Tower International,192180,Phoenix Tower International,2015-03-02 23:52:33.697922,2023-06-26 12:05:50.800659,27.95261,-82.36364,8326 East Adamo Drive,AMBPRO214,...,<NA>,<NA>,<NA>,"8326 East Adamo Drive , Tampa, FL 33619, UNITE...",NaN,No,Unconfirmed,Proximity Audit (0m),0.000000,0.0
2,1053602,Phoenix Tower International,<NA>,Phoenix Tower International,2022-10-26 15:53:59.431417,2023-06-26 12:05:50.800659,35.19402,-80.88992,4123 Barringer Drive,AMBPRO387,...,<NA>,<NA>,<NA>,"4123 Barringer Drive , Charlotte, NC 28217, UN...",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
3,1053602,Phoenix Tower International,192250,Phoenix Tower International,2015-03-02 23:52:33.804113,2023-06-26 12:05:50.800659,35.19402,-80.88992,4123 Barringer Drive,AMBPRO387,...,<NA>,<NA>,<NA>,"4123 Barringer Drive , Charlotte, NC 28217, UN...",NaN,No,Unconfirmed,Proximity Audit (1m),0.000000,0.0
4,1054266,Phoenix Tower International,<NA>,Phoenix Tower International,2022-10-26 16:35:12.716918,2023-06-26 12:05:50.800659,40.40552,-80.55032,1360 Cove Road,DDR604,...,<NA>,<NA>,<NA>,"1360 Cove Road , Weirton, WV 26062, UNITED STATES",NaN,No,Unconfirmed,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4444,1161490,SBA,118525,SBA,2015-02-03 20:22:46.964012,2017-11-08 10:40:17.274368,34.83366,-86.64344,Burwell Springs,AL40003-A,...,<NA>,<NA>,<NA>,", Huntsville, AL 35773, UNITED STATES",NaN,No,Active,Proximity Audit (0m),0.914725,0.9
4445,1161507,SBA,<NA>,SBA,2025-02-20 05:40:14.529083,2025-02-20 05:40:14.465301,33.03370,-85.77611,New Site 2 AL,AL40493-T,...,<NA>,<NA>,<NA>,"177 Tower Street , New Site, AL 36256, UNITED ...",NaN,No,Active,Focus Asset,NaN,NaN
4446,1161507,SBA,118581,SBA,2015-02-03 20:22:47.027416,2017-11-08 10:41:30.37165,33.03370,-85.77584,"New Site 2, AL",AL40493-T,...,<NA>,<NA>,<NA>,", New Site, AL 36256, UNITED STATES",NaN,No,Active,Proximity Audit (25m),25.222771,0.0
4447,1161508,SBA,<NA>,SBA,2025-02-20 05:40:16.728972,2025-02-20 05:40:16.66007,33.07055,-86.04802,Goodwater 2 AL,AL40494-T,...,<NA>,<NA>,<NA>,"22881 AL Hwy 9 , Goodwater, AL 35072, UNITED S...",NaN,No,Active,Focus Asset,NaN,NaN


In [47]:
case_5_3_candidates['focus_asset_id'].nunique()

2135

We got 2,135 groupings for this condition. Now let's add Tower Type as one of the matching condition.

In [48]:
def find_case_5_4_candidates(df_with_differences):

    case_5_4_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']
            ref_operator = reference_record['operator_name']
            ref_site_id = reference_record['operator_site_id']
            ref_status = reference_record['asset_status']
            ref_type = reference_record['type']

            if pd.isnull(ref_fcc) and pd.isnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'].isnull()) &
                    (paired_records_df['faa_study_number'].isnull()) &
                    (paired_records_df['operator_name'] == ref_operator) &
                    (paired_records_df['operator_site_id'] == ref_site_id) &
                    (paired_records_df['asset_status'] == ref_status) &
                    (paired_records_df['type'] == ref_type)
                ]

                if not matching_paired_records.empty:
                    case_5_4_records.append(reference_record_df)
                    case_5_4_records.append(matching_paired_records)

    if case_5_4_records:
        case_5_4_candidates = pd.concat(case_5_4_records, ignore_index=True)
        return case_5_4_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_5_4_candidates = find_case_5_4_candidates(df_with_differences)
case_5_4_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1056851,sgaither,<NA>,sgaither,2015-03-02 23:52:48.491158,2022-12-27 17:28:52.637154,32.99694,-96.90873,3044 Old Denton Road 3,NT1317,...,<NA>,<NA>,<NA>,"3044 Old Denton Road 3 , Carrollton, TX 75007,...",NaN,No,Speculative,Focus Asset,NaN,NaN
1,1056851,sgaither,1056851,sgaither,2015-03-02 23:52:48.491158,2022-12-27 17:28:52.637154,32.99694,-96.90873,3044 Old Denton Road 3,NT1317,...,<NA>,<NA>,<NA>,"3044 Old Denton Road 3 , Carrollton, TX 75007,...",NaN,No,Speculative,Manual Edit,0.000000,NaN
2,1058500,sgaither,<NA>,sgaither,2015-03-02 23:52:35.640475,2024-11-01 16:31:15.438971,32.28873,-110.97393,4500 N Oracle Rd,KRO2924,...,<NA>,<NA>,<NA>,"4500 N Oracle Rd , Tucson, AZ 85705, UNITED ST...",NaN,No,Active,Focus Asset,NaN,NaN
3,1058500,sgaither,1058500,sgaither,2015-03-02 23:52:35.640475,2024-11-01 16:31:15.438971,32.28873,-110.97393,4500 N Oracle Rd,KRO2924,...,<NA>,<NA>,<NA>,"4500 N Oracle Rd , Tucson, AZ 85705, UNITED ST...",NaN,No,Active,Multiple matches,0.000000,NaN
4,1058559,sgaither,<NA>,sgaither,2015-03-02 23:52:38.664205,2024-12-13 17:32:37.148467,26.48384,-98.38314,36868 Fm 490,UVL002,...,<NA>,<NA>,<NA>,"36868 Fm 490 , Edinburg, TX 78541, UNITED STATES",NaN,No,Speculative,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4076,1161490,SBA,118525,SBA,2015-02-03 20:22:46.964012,2017-11-08 10:40:17.274368,34.83366,-86.64344,Burwell Springs,AL40003-A,...,<NA>,<NA>,<NA>,", Huntsville, AL 35773, UNITED STATES",NaN,No,Active,Proximity Audit (0m),0.914725,0.9
4077,1161507,SBA,<NA>,SBA,2025-02-20 05:40:14.529083,2025-02-20 05:40:14.465301,33.03370,-85.77611,New Site 2 AL,AL40493-T,...,<NA>,<NA>,<NA>,"177 Tower Street , New Site, AL 36256, UNITED ...",NaN,No,Active,Focus Asset,NaN,NaN
4078,1161507,SBA,118581,SBA,2015-02-03 20:22:47.027416,2017-11-08 10:41:30.37165,33.03370,-85.77584,"New Site 2, AL",AL40493-T,...,<NA>,<NA>,<NA>,", New Site, AL 36256, UNITED STATES",NaN,No,Active,Proximity Audit (25m),25.222771,0.0
4079,1161508,SBA,<NA>,SBA,2025-02-20 05:40:16.728972,2025-02-20 05:40:16.66007,33.07055,-86.04802,Goodwater 2 AL,AL40494-T,...,<NA>,<NA>,<NA>,"22881 AL Hwy 9 , Goodwater, AL 35072, UNITED S...",NaN,No,Active,Focus Asset,NaN,NaN


In [49]:
case_5_4_candidates['focus_asset_id'].nunique()

1990

We got 1,990 groupings for this condition. Now lastly let's add Tower name as one of the matching condition.

In [50]:
def find_case_5_5_candidates(df_with_differences):

    case_5_5_records = []

    grouped = df_with_differences.groupby('focus_asset_id')

    for focus_asset_id, group in grouped:
        reference_record_df = group[group['associated_asset_id'].isnull()]
        paired_records_df = group[group['associated_asset_id'].notnull()]

        if len(reference_record_df) == 1:
            reference_record = reference_record_df.iloc[0]
            ref_fcc = reference_record['fcc_asr_number']
            ref_faa = reference_record['faa_study_number']
            ref_operator = reference_record['operator_name']
            ref_site_id = reference_record['operator_site_id']
            ref_status = reference_record['asset_status']
            ref_type = reference_record['type']
            ref_tower_name = reference_record['name']

            if pd.isnull(ref_fcc) and pd.isnull(ref_faa):
                matching_paired_records = paired_records_df[
                    (paired_records_df['fcc_asr_number'].isnull()) &
                    (paired_records_df['faa_study_number'].isnull()) &
                    (paired_records_df['operator_name'] == ref_operator) &
                    (paired_records_df['operator_site_id'] == ref_site_id) &
                    (paired_records_df['asset_status'] == ref_status) &
                    (paired_records_df['type'] == ref_type) &
                    (paired_records_df['name'] == ref_tower_name)
                ]

                if not matching_paired_records.empty:
                    case_5_5_records.append(reference_record_df)
                    case_5_5_records.append(matching_paired_records)

    if case_5_5_records:
        case_5_5_candidates = pd.concat(case_5_5_records, ignore_index=True)
        return case_5_5_candidates.drop_duplicates()
    else:
        return pd.DataFrame(columns=df_with_differences.columns)

case_5_5_candidates = find_case_5_5_candidates(df_with_differences)
case_5_5_candidates

,focus_asset_id,focus_asset,associated_asset_id,source,created_at,updated_at,latitude,longitude,name,operator_site_id,...,faa_study_number,cdbs_facility_id,region,address,construction_date,stealth,asset_status,audit_reason,distance_to_reference,agldiff_to_reference
0,1056851,sgaither,<NA>,sgaither,2015-03-02 23:52:48.491158,2022-12-27 17:28:52.637154,32.99694,-96.90873,3044 Old Denton Road 3,NT1317,...,<NA>,<NA>,<NA>,"3044 Old Denton Road 3 , Carrollton, TX 75007,...",NaN,No,Speculative,Focus Asset,NaN,NaN
1,1056851,sgaither,1056851,sgaither,2015-03-02 23:52:48.491158,2022-12-27 17:28:52.637154,32.99694,-96.90873,3044 Old Denton Road 3,NT1317,...,<NA>,<NA>,<NA>,"3044 Old Denton Road 3 , Carrollton, TX 75007,...",NaN,No,Speculative,Manual Edit,0.000000,NaN
2,1058500,sgaither,<NA>,sgaither,2015-03-02 23:52:35.640475,2024-11-01 16:31:15.438971,32.28873,-110.97393,4500 N Oracle Rd,KRO2924,...,<NA>,<NA>,<NA>,"4500 N Oracle Rd , Tucson, AZ 85705, UNITED ST...",NaN,No,Active,Focus Asset,NaN,NaN
3,1058500,sgaither,1058500,sgaither,2015-03-02 23:52:35.640475,2024-11-01 16:31:15.438971,32.28873,-110.97393,4500 N Oracle Rd,KRO2924,...,<NA>,<NA>,<NA>,"4500 N Oracle Rd , Tucson, AZ 85705, UNITED ST...",NaN,No,Active,Multiple matches,0.000000,NaN
4,1058559,sgaither,<NA>,sgaither,2015-03-02 23:52:38.664205,2024-12-13 17:32:37.148467,26.48384,-98.38314,36868 Fm 490,UVL002,...,<NA>,<NA>,<NA>,"36868 Fm 490 , Edinburg, TX 78541, UNITED STATES",NaN,No,Speculative,Focus Asset,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3163,1161425,SBA,463695,SBA,2015-10-07 14:53:10.478301,2017-11-08 10:42:49.585368,30.61561,-88.13534,Rivere du Chein,AL17028-A,...,<NA>,<NA>,<NA>,", Mobile, AL 36693, UNITED STATES",NaN,No,Active,Proximity Audit (1m),1.108629,0.0
3164,1161426,SBA,<NA>,SBA,2025-02-20 05:37:04.504778,2025-02-20 05:37:04.442489,30.79995,-88.08447,Saraland,AL17029-A,...,<NA>,<NA>,<NA>,"350 Industrial Parkway , Saraland, AL 36571, U...",NaN,No,Active,Focus Asset,NaN,NaN
3165,1161426,SBA,463696,SBA,2015-10-07 14:53:10.478301,2017-11-08 10:42:33.923615,30.79996,-88.08448,Saraland,AL17029-A,...,<NA>,<NA>,<NA>,", Saraland, AL 36571, UNITED STATES",NaN,No,Active,Proximity Audit (1m),1.464594,0.0
3166,1161490,SBA,<NA>,SBA,2025-02-20 05:39:36.736587,2025-02-20 05:39:36.668613,34.83366,-86.64343,Burwell Springs,AL40003-A,...,<NA>,<NA>,<NA>,"7485 Pulaski Pike , Huntsville, AL 35773, UNIT...",NaN,No,Active,Focus Asset,NaN,NaN


In [51]:
case_5_5_candidates['focus_asset_id'].nunique()

1541

In [52]:
case_5_candidates.to_csv('case_5_candidates.csv')
case_5_1_candidates.to_csv('case_5_1_candidates.csv')
case_5_2_candidates.to_csv('case_5_2_candidates.csv')
case_5_3_candidates.to_csv('case_5_3_candidates.csv')
case_5_4_candidates.to_csv('case_5_4_candidates.csv')
case_5_5_candidates.to_csv('case_5_5_candidates.csv')

We got 1,541 groupings for this condition.

Thoughts: 

1.) Cases 1 - 3 will come more into play when we will be loading more new sheets from the company as these cases are highly related scope to new sheets. Note that we didn't had an extensive analysis for these three cases, unlike cases 4 and 5 where we add parameters in the matching condition one-at-a-time to quantify the surviving groupings, because Cases 1 - 3 still has the presence of FCC ASR and FAA registry number. BUT, that doesn't mean that we will not be exploring more for these three cases, we will after this.

2.) Case 4 is promising. But, it will not totally diminish the number of groupings in the prox audits. What most likely it will do is just do an automerge for each grouping that conforms to case 4's logic. That doesn't mean that the whole group will be away, because there will be instances where there are more than 2 records in a group. Say we have 5 records -- if one of the paired records and the reference record will be automerging, there will still be 3 more remaining records left untouched. Thus, the new group will now have 4 records. But at least this will be an easy "confirm correct and continue" by the analyst, most of the time. 

3.) Case 5 will come into play for cases where the data reported by the company is "not so usable" since there are no government identifier being reported for these tower assets (i.e., FCC ASR and FAA). Probably, these are tower assets that are not registered to any government entity.

4.) Also, we will explore more the outputs we have for cases 4 and 5 to design matching criteria that will give us more accurate results.